projects/python-client/src/datapane/client/api/ipython_utils.py

```diff
@@ -19,10 +25,219 @@ def block_to_iframe(block: BaseElement) -> str:
     """
-    # importing within function scope to avoid circular dependency
     from .report.core import App
 
     app = App(block)
     block_html_string = app.stringify(template_name="ipython_template.html")
 
     return block_html_string
+
+
+@capture_event("Get Jupyter Notebook JSON")
+def get_jupyter_notebook_json() -> dict:
+    """Get the JSON for the current Jupyter notebook
+
+    Returns:
+        Notebook JSON
+    """
+    import ipynbname
+
+    try:
+        nb_path = ipynbname.path()
+        notebook_json = json.loads(open(nb_path, encoding="utf-8").read())
+    except FileNotFoundError as e:
+        raise DPError("Notebook not found. This command must be executed from within a notebook environment.") from e
+
+    return notebook_json
+
+
+@capture_event("Get VSCode Notebook JSON")
+def get_vscode_notebook_json() -> dict:
+    """Get the JSON for the current VSCode notebook
+
+    Returns:
+        Notebook JSON
+    """
+    import ipynbname
+
+    try:
+        nb_path = str(ipynbname.path())
+
+        # VSCode path name in sesssion is suffixed with -jvsc-[identifier]
+        nb_path = nb_path.split("-jvsc-")[0] + ".ipynb"
+
+        notebook_json = json.loads(open(nb_path, encoding="utf-8").read())
+    except FileNotFoundError as e:
+        raise DPError("Notebook not found. This command must be executed from within a notebook environment.") from e
+
+    return notebook_json
+
+
+@capture_event("Get Colab Notebook JSON")
+def get_colab_notebook_json() -> dict:
+    """Get the JSON for the current Colab notebook
+
+    Returns:
+        Notebook JSON
+    """
+    import ipynbname
+    from google.colab import auth
+    from googleapiclient.discovery import build
+    from googleapiclient.http import MediaIoBaseDownload
+
+    # Get the notebook's Google Drive file_id
+    file_id = ipynbname.name().replace("fileId=", "")
+
+    try:
+        auth.authenticate_user()
+    except Exception as e:
+        raise DPError(
+            "Google Drive authentication failed. Please allow this notebook to access your Google Drive."
+        ) from e
+
+    drive_service = build("drive", "v3")
+
+    request = drive_service.files().get_media(fileId=file_id)
+    downloaded = io.BytesIO()
+    downloader = MediaIoBaseDownload(downloaded, request)
+    done = False
+    while done is False:
+        # _ is a placeholder for a progress object that we ignore.
+        # (Our file is small, so we skip reporting progress.)
+        _, done = downloader.next_chunk()
+
+    downloaded.seek(0)
+    notebook_json = json.loads(downloaded.read().decode("utf-8"))
+
+    return notebook_json
+
+
+def get_notebook_json() -> dict:
+    """Get the JSON for the current Jupyter, Colab, or VSCode notebook
+
+    Returns:
+        Notebook JSON
+    """
+    if "COLAB_GPU" in os.environ:
+        notebook_json = get_colab_notebook_json()
+    elif "VSCODE_PID" in os.environ:
+        notebook_json = get_vscode_notebook_json()
+    else:
```

Comment:
is there a way to detect normal jupyter(lab) here in another `elif`, so that if that falls through we go to the last `else` as follows,

```python
else:
   raise DPError("Can't detect notebook environment")
```